### PR TITLE
Final Preparations for 330 - Phu

### PR DESCRIPTION
**Pull request by phudtran**

### CODE REVIEW

File: docs/user/getting_started.md [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+Simply run the script below in the Mizar directory.
+
+```
+$ ./kind_setup.sh
+```
+
+This script does the following:
+
+* Create a multi-node local Kubernetes cluster with Kind.
+* Build all of the necessary Docker images.
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: add a brief intro of each docker images' purpose would be great

File: docs/user/getting_started.md [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+If you are testing out Mizar with Kind, a script has been included to setup a local Kubernetes cluster, and install all of the components needed to run Mizar.
+Simply run the script below in the Mizar directory.
+
+```
+$ ./kind_setup.sh
+```
+
+This script does the following:
+
+* Create a multi-node local Kubernetes cluster with Kind.
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: [ADD] this kubernetes cluster is used as a cluster of VM's or droplets, and will later be used to provision test environment

File: docs/user/getting_started.md [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+* Build all of the necessary Docker images.
+* Apply all of the Mizar CRDs.
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: what is CRD. maybe provide a link to the document on what is CRD here

File: docs/user/getting_started.md [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+* Deploy the Mizar Operators
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: What does the operators do?

File: docs/user/getting_started.md [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+    metadata:
+      labels:
+        app: nginx
+    spec:
+      containers:
+        - name: nginx
+          image: nginx:1.7.9
+          ports:
+            - containerPort: 80
+```
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: You can run ping test from a Nginx machine: 

kubectl exec -it nginx-deployment-your-nigix-machine /bin/bash 

Now, you are running in interactive mode inside the nginx docker container. From the docker container, ping a different host: 

ping 10.0.0.5

File: docs/user/getting_started.md [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+apiVersion: mizar.com/v1
+kind: Divider
+metadata:
+  name: d1
+spec:
+  vpc: "vpc1"
+  status: "Init"
+```
+Note: To deploy, simply copy this to a file and run
+```kubectl deploy file_name_here.yaml```
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: Typo: kubectl create -f file_name_here.yaml

File: docs/user/getting_started.md [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+kind: Bouncer
+metadata:
+  name: b1
+spec:
+  vpc: "vpc1"
+  net: "net1"
+  status: "Init"
+```
+Note: To deploy, simply copy this to a file and run
+```kubectl deploy file_name_here.yaml```
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: typo: kubectl create -f file_name_here.yaml

File: docs/user/getting_started.md [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+  name: vpc2
+spec:
+  vni: "1"
+  ip: "20.0.0.0"
+  prefix: "16"
+  dividers: 10
+  status: "Init"
+```
+Note: To deploy, simply copy this to a file and run
+```kubectl deploy file_name_here.yaml```
```
Comment (clu2xlu) [inline docs/user/getting_started.md]: Typo: kubectl create -f file_name_here.yaml

File: docs/user/getting_started.md [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+
+### Mizar Compilation Prerequisites
+Make is used to build the entire data-plane code. An Ubuntu-based host is recommended for development and functionality testing. To compile, run Mizar on an existing host, please follow the following steps:
+
+* Clone Mizar
+```
+$ git clone --recurse-submodules -j8 https://github.com/futurewei-cloud/mizar.git
+```
+* Enter the mizar directory and run the bootstrap.sh script.
+    * This script will the neccessary components to compile Mizar, and run it's unit tests. These include
```
[the remark this answers is not in the document]
Comment (phudtran) [reply docs/user/getting_started.md]: Fixed.

File: docs/user/getting_started.md [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+        * Python
+        * Docker
+        * Build Essentials
+        * Cmocka
+        * Clang-7
+        * Llvm-7
```
[the remark this answers is not in the document]
Comment (phudtran) [reply docs/user/getting_started.md]: I've added a link to the official site for these packages.

File: docs/user/getting_started.md [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+After completing the prerequisites above, you can now compile mizar. Enter the mizar directory and run the command below.
+
+```
+$ make
+```
+
+* You can now optionally run the unit tests and coverage reports by running the commands below.
+
+```
+$ run_unittests
```
[the remark this answers is not in the document]
Comment (phudtran) [reply docs/user/getting_started.md]: Fixed.

File: docs/user/getting_started.md [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+$ make lcov
```
[the remark this answers is not in the document]
Comment (phudtran) [reply docs/user/getting_started.md]: This is explained in the bullet point preceding these commands.

File: docs/user/getting_started.md [hunk 13]
```diff
@@ -0,0 +1,216 @@
...
+```
+
+### New Kind Cluster
+
+If you are testing out Mizar with Kind, a script has been included to setup a local Kubernetes cluster, and install all of the components needed to run Mizar.
```
[the remark this answers is not in the document]
Comment (phudtran) [reply docs/user/getting_started.md]: Great idea. Added a note for the user to validate their installations before proceeding with the setup script.

File: docs/user/getting_started.md (hunk 3 above)
Comment (phudtran) [reply docs/user/getting_started.md]: I added a link to the official CRD documentation

File: docs/user/getting_started.md (hunk 4 above)
Comment (phudtran) [reply docs/user/getting_started.md]: This is explained in the design portion of the doc.

File: docs/user/getting_started.md (hunk 5 above)
Comment (phudtran) [reply docs/user/getting_started.md]: I can add this to be beginner friendly.

File: docs/user/getting_started.md (hunk 1 above)
Comment (phudtran) [reply docs/user/getting_started.md]: Great idea, I'll add a brief note on what each docker image is running.

File: docs/user/getting_started.md [hunk 14]
```diff
@@ -0,0 +1,216 @@
...
+Simply run the script below in the Mizar directory.
+
+```
+$ ./kind_setup.sh
+```
+
```
[the remark this answers is not in the document]
Comment (phudtran) [reply docs/user/getting_started.md]: The errors you see are from kubectl doing a cleanup before deploying the operator and daemon.
The operator and daemon are never there when a new cluster is setup so there will always be an error.
I will suppress these false errors in the scripts in a later commit.

Will definitely add a note explaining to the user how to check the readiness/status of the resources!

File: docs/design/management_plane.md [hunk 15]
```diff
@@ -291,14 +291,30 @@ For example:
 9. The Endpoint Operator sees that the Endpoint's state is now provisioned.
 10. The Endpoint Operator updates its own state store with the newly provisioned Endpoint.
 
+### Scaled Endpoint Object Create Workflow
+![Scaled_Endpoint](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/futurewei-cloud/mizar/dev-next/docs/design/puml/create_scaled_endpoint_workflow.puml)
+
+1. The Endpoint Operator is notified that a Scaled Endpoint has been created
+2. The Endpoint Operator allocates an IP, mac, etc, for the new Scaled Endpoint Object
+3. CNI Service is informed about the newly created Scaled Endpoint
+4. The CNI Services call backs to the CNI about the new Scaled Endpoint
```
Comment (zasherif) [inline docs/design/management_plane.md]: 3 and 4 are only for simple endpoints.

File: docs/design/management_plane.md [hunk 16]
```diff
@@ -291,14 +291,30 @@ For example:
...
+5. The Scaled Endpoint object is updated with Endpoint Objects as its backends
```
Comment (zasherif) [inline docs/design/management_plane.md]: We will need a separate workflow description for backend update.

File: docs/design/management_plane.md [hunk 17]
```diff
@@ -336,15 +352,82 @@ For example:
...
+6. The Bouncer Operator deletes the Endpoint information from all Bouncers.
+7. The Bouncer Operator deletes all Bouncer information from the Endpoint.
+8. The Bouncer Operator calls back to the Endpoint Operator.
+9. The Endpoint Operator deletes the Endpoint Object from its cache.
+
+### Scaled Endpoint Object Delete Workflow
+![Scaled_Endpoint](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.githubusercontent.com/futurewei-cloud/mizar/dev-next/docs/design/puml/delete_scaled_endpoint_workflow.puml)
+
+1. The CNI Services sees the Scaled Endpoint Delete request.
+2. The CNI service deletes the Scaled Endpoint Object.
```
Comment (zasherif) [inline docs/design/management_plane.md]: Remove these. The CNI is not involved in the scaled endpoint delete. We have a built-ins operator that watches for services/pods. We will need to add a workflow for the builtin operator.

File: docs/design/puml/create_scaled_endpoint_workflow.puml [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+
+database API_Server
+collections CNI_Service
+collections Bouncer_Operator
+collections Endpoints_Operator
+
+==Scaled Endpoint Created==
+Endpoints_Operator --> API_Server: Sees endpoint created
+Endpoints_Operator -> Endpoints_Operator: Allocates IP, Mac, etc.
+Endpoints_Operator -> CNI_Service: On Endpoint-Alloc
```
Comment (zasherif) [inline docs/design/puml/create_scaled_endpoint_workflow.puml]: The CNI service is not involved.

File: docs/design/puml/create_scaled_endpoint_workflow.puml [hunk 19]
```diff
@@ -0,0 +1,27 @@
...
+CNI_Service -> CNI_Service: Write to CNI file.
```
Comment (zasherif) [inline docs/design/puml/create_scaled_endpoint_workflow.puml]: ditto

File: docs/design/puml/delete_endpoint_workflow.puml [hunk 20]
```diff
@@ -10,21 +10,17 @@ collections Bouncer_Operator
 collections Endpoints_Operator
 
 ==Endpoint Delete Requested==
-rnote over API_Server: == Endpoint Status: TBD
-Network_Operator --> API_Server: Sees Endpoint status TBD
+CNI_Service --> API_Server: Sees Endpoint delete request
+CNI_Service -> API_Server: Delete endpoint
```
Comment (zasherif) [inline docs/design/puml/delete_endpoint_workflow.puml]: ditto

File: docs/design/puml/delete_scaled_endpoint_workflow.puml [hunk 21]
```diff
@@ -0,0 +1,25 @@
...
+autonumber
+
+database API_Server
+collections CNI_Service
+collections Bouncer_Operator
+collections Endpoints_Operator
+
+==Scaled Endpoint Delete Requested==
+CNI_Service --> API_Server: Sees Endpoint delete request
+CNI_Service -> API_Server: Delete endpoint
```
Comment (zasherif) [inline docs/design/puml/delete_scaled_endpoint_workflow.puml]: ditto

File: docs/user/getting_started.md [hunk 22]
```diff
@@ -0,0 +1,252 @@
...
+    * There is also an optional kernel update included in the script if you wish to update to a compatible version
+
+### Linux Kernel Update
+
+For TCP to function properly, you will need to update your Kernel version to at least 5.6-rc2 on every node. A script, ```kernel_update.sh``` is provided in the Mizar repo to download and update your machine's kernel if you do not wish to build the kernel source code yourself.
+
+
+## Setup
+
+### Compiling Mizar
```
Comment (zasherif) [inline docs/user/getting_started.md]: Shall we move the compilation and test running to the developer guide?

File: docs/user/getting_started.md [hunk 23]
```diff
@@ -0,0 +1,252 @@
...
+$ make
+```
+
+* You can now optionally run the unit tests and coverage reports by running the commands below.
+
+```
+$ make run_unittests
+$ make lcov
+```
+
```
Comment (zasherif) [inline docs/user/getting_started.md]: Move to dev guide

File: docs/user/getting_started.md [hunk 24]
```diff
@@ -0,0 +1,252 @@
...
+This script does the following:
+
+* Create a multi-node local Kubernetes cluster with Kind.
+* Build the Kind-Node, Mizar-Daemon, and Mizar-Operator docker images
+* Apply all of the Mizar [CRDs](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/).
+* Deploy the Mizar [Operators](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/).
+* Deploy the Mizar Daemon
+* Install the Mizar CNI Plugin
```
Comment (zasherif) [inline docs/user/getting_started.md]: Reorganize this, move "New kind cluster" to "Kind (Kubernetes in Docker)"

File: docs/user/getting_started.md [hunk 25]
```diff
@@ -0,0 +1,252 @@
...
+
+### Existing Kubernetes Cluster
+
+If you are installing Mizar on an existing Kubernetes cluster, please complete the following steps.
+
+To be expanded...
```
Comment (zasherif) [inline docs/user/getting_started.md]: Do we need to mention that? If you have kubectl running with a cluster, we are pretty much done

File: docs/user/getting_started.md [hunk 26]
```diff
@@ -0,0 +1,252 @@
...
+  vpc: "vpc1"
+  net: "net1"
+  status: "Init"
+```
+Note: To deploy, simply copy this to a file and run
+```
+$ kubectl create -f file_name_here.yaml
+```
+
+This Bouncer will be created for the Network with name "net1" that is part of the VPC with name "vpc1"
```
Comment (zasherif) [inline docs/user/getting_started.md]: This part should be moved to the developer guide, or not to be mentioned all together. Scaling up the number of dividers/bouncers shall be done only by increasing the dividers/bouncers count in the VPC/Net yaml file. Dividers and Bouncers shall not be created directly like this...

File: mgmt/dp/mizar/workflows/nets/delete.py [hunk 27]
```diff
@@ -19,9 +21,14 @@ def run(self):
 		logger.info("Run {task}".format(task=self.__class__.__name__))
 		n = nets_opr.store.get_net(self.param.name)
 		n.set_obj_spec(self.param.spec)
+
```
Comment (zasherif) [inline mgmt/dp/mizar/workflows/nets/delete.py]: empty line

File: mgmt/dp/mizar/workflows/nets/delete.py [hunk 28]
```diff
@@ -19,9 +21,14 @@ def run(self):
 		logger.info("Run {task}".format(task=self.__class__.__name__))
 		n = nets_opr.store.get_net(self.param.name)
 		n.set_obj_spec(self.param.spec)
+		
+		while len(endpoints_opr.store.get_eps_in_net(n.name)):
+			pass
+
```
Comment (zasherif) [inline mgmt/dp/mizar/workflows/nets/delete.py]: ditto

File: mgmt/dp/mizar/workflows/nets/delete.py [hunk 29]
```diff
@@ -19,9 +21,14 @@ def run(self):
 		logger.info("Run {task}".format(task=self.__class__.__name__))
 		n = nets_opr.store.get_net(self.param.name)
 		n.set_obj_spec(self.param.spec)
+		
+		while len(endpoints_opr.store.get_eps_in_net(n.name)):
+			pass
```
Comment (zasherif) [inline mgmt/dp/mizar/workflows/nets/delete.py]: Add a todo commit here and refer to the error handling status...

File: mgmt/dp/mizar/workflows/vpcs/delete.py [hunk 30]
```diff
@@ -15,7 +17,10 @@ def run(self):
         logger.info("Run {task}".format(task=self.__class__.__name__))
         v = vpcs_opr.store.get_vpc(self.param.name)
         v.set_obj_spec(self.param.spec)
+        while len(nets_opr.store.get_nets_in_vpc(v.name)):
+            pass
```
Comment (zasherif) [inline mgmt/dp/mizar/workflows/vpcs/delete.py]: ditto: TODO to error handling status

File: mgmt/etc/luigi.cfg [hunk 31]
```diff
@@ -0,0 +1,6 @@
+
+[core]
+no_configure_logging=true
+
+[worker]
+no_install_shutdown_handler=true
```
Comment (zasherif) [inline mgmt/etc/luigi.cfg]: add a new line :/ Sorry

File: mgmt/test/func_test/basic/yaml/test_net.yaml [hunk 32]
```diff
@@ -6,6 +6,6 @@ spec:
   vni: "1"
   ip: "20.0.20.0"
   prefix: "24"
-  bouncers: 2
+  bouncers: 10
   vpc: "vpc2"
   status: "Init"
```
Comment (zasherif) [inline mgmt/test/func_test/basic/yaml/test_net.yaml]: new line

File: mgmt/test/func_test/basic/yaml/test_vpc.yaml [hunk 33]
```diff
@@ -6,5 +6,5 @@ spec:
   vni: "1"
   ip: "20.0.0.0"
   prefix: "16"
-  dividers: 2
+  dividers: 10
   status: "Init"
```
Comment (zasherif) [inline mgmt/test/func_test/basic/yaml/test_vpc.yaml]: new line

File: docs/design/management_plane.md (hunk 15 above)
Comment (phudtran) [reply docs/design/management_plane.md]: Gotcha, Fixed.

File: docs/design/management_plane.md (hunk 16 above)
Comment (phudtran) [reply docs/design/management_plane.md]: I have split the workflows. One for create scaled endpoint, and one for update scaled endpoint

File: docs/design/management_plane.md (hunk 17 above)
Comment (phudtran) [reply docs/design/management_plane.md]: Gotcha, Fixed.

File: docs/design/puml/create_scaled_endpoint_workflow.puml (hunk 18 above)
Comment (phudtran) [reply docs/design/puml/create_scaled_endpoint_workflow.puml]: Gotcha, Fixed.

File: docs/design/puml/create_scaled_endpoint_workflow.puml (hunk 19 above)
Comment (phudtran) [reply docs/design/puml/create_scaled_endpoint_workflow.puml]: Fixed.

File: docs/design/puml/delete_endpoint_workflow.puml (hunk 20 above)
Comment (phudtran) [reply docs/design/puml/delete_endpoint_workflow.puml]: This is just for the regular endpoint.

File: docs/design/puml/delete_scaled_endpoint_workflow.puml (hunk 21 above)
Comment (phudtran) [reply docs/design/puml/delete_scaled_endpoint_workflow.puml]: Fixed.

File: docs/user/getting_started.md (hunk 22 above)
Comment (phudtran) [reply docs/user/getting_started.md]: Sure, I've moved it to the developer guide.

File: docs/user/getting_started.md (hunk 23 above)
Comment (phudtran) [reply docs/user/getting_started.md]: moved.

File: docs/user/getting_started.md (hunk 24 above)
Comment (phudtran) [reply docs/user/getting_started.md]: moved.

File: docs/user/getting_started.md (hunk 25 above)
Comment (phudtran) [reply docs/user/getting_started.md]: I'll remove this for now. I think there are couple of things that we rely on with Kind that may not be apparent when compared to a regular Kubernetes cluster. For example, with Kind, we have node images that share directories with our local machine. (The "extraMounts" parameter).

File: docs/user/getting_started.md (hunk 26 above)
Comment (phudtran) [reply docs/user/getting_started.md]: Ok, I'll remove this.

File: mgmt/dp/mizar/workflows/nets/delete.py (hunk 27 above)
Comment (phudtran) [reply mgmt/dp/mizar/workflows/nets/delete.py]: Fixed

File: mgmt/dp/mizar/workflows/nets/delete.py (hunk 28 above)
Comment (phudtran) [reply mgmt/dp/mizar/workflows/nets/delete.py]: Fixed

File: mgmt/dp/mizar/workflows/nets/delete.py (hunk 29 above)
Comment (phudtran) [reply mgmt/dp/mizar/workflows/nets/delete.py]: Added

File: mgmt/dp/mizar/workflows/vpcs/delete.py (hunk 30 above)
Comment (phudtran) [reply mgmt/dp/mizar/workflows/vpcs/delete.py]: Added

File: mgmt/etc/luigi.cfg (hunk 31 above)
Comment (phudtran) [reply mgmt/etc/luigi.cfg]: Fixed.

File: mgmt/test/func_test/basic/yaml/test_net.yaml (hunk 32 above)
Comment (phudtran) [reply mgmt/test/func_test/basic/yaml/test_net.yaml]: Fixed.

File: mgmt/test/func_test/basic/yaml/test_vpc.yaml (hunk 33 above)
Comment (phudtran) [reply mgmt/test/func_test/basic/yaml/test_vpc.yaml]: Fixed

**Pull request merged**